src/inmanta/execute/scheduler.py

```diff
@@ -256,23 +256,24 @@ def define_types(self, compiler: "Compiler", statements: Sequence["Statement"],
             implement.evaluate()
 
         compiler.plugins = {k: v for k, v in types_and_impl.items() if isinstance(v, plugins.Plugin)}
-        types = {k: v for k, v in types_and_impl.items() if isinstance(v, Type)}
-
-        # normalize plugins
-        for p in compiler.plugins.values():
-            p.normalize()
+        self.types = {k: v for k, v in types_and_impl.items() if isinstance(v, Type)}
 
-        # give type info to all types, to normalize blocks inside them
-        # normalize implementations last because they have subblocks that might depend on other type information
-        for t in sorted(types.values(), key=lambda t: isinstance(t, Implementation)):
+        # give type info to all types (includes plugins), to normalize blocks inside them
+        for t in sorted(
+            self.types.values(),
+            key=lambda t: (
+                # normalize implementations last because they have subblocks that might depend on other type information
+                isinstance(t, Implementation),
+                # normalize entities second last because it may call validate on its attribute types to validate defaults
+                isinstance(t, Entity),
+            ),
+        ):
```

Comment:
If I understand correctly, this implementation will normalize plugins first since `key = (0, 0)` for all of them.
1. Does it matter that they are normalized before `Implementations` and `Entities` ?
2. Does the relative order in which plugins are normalized matter ? Do we need it to be deterministic ? If I understand correctly this order depends on the order in which the plugin files are loaded and the plugins functions registered.

Reply:
Indeed, that's how it works. With the two boolean sort keys, it gets right to the edge of being too complex a solution for me. But I still (barely) prefer it over a more verbose one. What do you think?

1. It has to be before both implementations and entities, because plugin normalization is standalone, while the other two may need other type information. The fact that this is so is a bit of on implementation detail that is assumed to hold here. I don't love it but I think it's the best we can do for now.
2. Doesn't matter. Deterministic in the sense that two consecutive compiles do it in the same order is important for deterministic error reporting etc. Deterministic in the sense that this implementation maintains the same order as the one on master isn't that important.

Reply:
> Indeed, that's how it works. With the two boolean sort keys, it gets right to the edge of being too complex a solution for me. But I still (barely) prefer it over a more verbose one. What do you think?

No strong opinion. I haven't fully considered performance but couldn't we add a class variable `normalization_priority` and sort on this. This part would be more readable and the ordering explicit

Reply:
I considered it, but I felt that it shifts the problem more than solves it. Then suddenly every class has to decide what a sane priority is, which means it has to know the priorities of other type classes.